### ISSUE TITLE
Some combinations of UGens can cause downstreams UGens to be incorrectly removed from the SynthDef

### ISSUE THREAD

**Issue by scztt** [supercollider/supercollider#972]
A relatively simple example:

```supercollider
d = SynthDef(\spreaderBroke, { 
    arg steerDeg = 0.3;
    var steer, angB;

    steer = (steerDeg * 1.1).neg;
    steer.poll(label: \steer);
    angB = (steer.neg + 1.2).neg;
    Out.kr(0, steer);
}).add;
```

In this case, both the poll and the Out.kr, as well as some of the binop UGens, do not appear in the final SynthDef.

This issue occurs during the optimization stage - some combination of binary multiplication/add/subtract operations hits optimization in such a way that the resulting UGens don't seem to have the correct descendants, and are eventually optimized right out of the graph.

This is quite serious - the above ugen combinations (and possibly others) can potentially nullify any dependent ugens downstream. :(

This is happening somewhere in the SynthDef:finishBuild stage, and may be related to the BinaryOpUGen:optimize\* methods. This is a 3.6 regression, and the big changes to SynthDef optimization in those methods is a good candidate.

I've been looking on a fix, but it's a tough one. If someone else can make better progress, by all means.... otherwise, for lack of better options, maybe revert above mentioned BinaryOpUGen optimizations, to at least fix the bug?

**Comment by scztt** [comment]
@timblechmann - Can you take a look at this? I believe it was caused by one of your changes to BasicOpsUGen.sc between a63404d and 19a7217 (I remember bisecting was difficult because of some interdependencies, but I can try again if it would help). If there's not a clear fix for this, I'm not sure what else to do other than revert the optimizations to ensure we're at least generating the graph correctly.

**Comment by telephon** [comment]
Just for the record:
In this specific case, it is triggered by the return of ````in```` instead of the ugen:

````
Poll : UGen {
	*ar { arg trig, in, label, trigid = -1;
		this.multiNewList(['audio', trig, in, label, trigid]);
		^in;
	}
	*kr { arg trig, in, label, trigid = -1;
		this.multiNewList(['control', trig, in, label, trigid]);
		^in;
	}
	*new { arg trig, in, label, trigid = -1;
		var rate = in.asArray.collect(_.rate).unbubble;
		this.multiNewList([rate, trig, in, label, trigid]);
		^in;
	}
````

The method that removes the poll ugen is:

````optimizeAddNeg````

````optimizeAddNeg {
		var a, b;
		#a, b = inputs;

		if (b.isKindOf(UnaryOpUGen) and: { b.operator == 'neg' }) {
			// a + b.neg -> a - b
			if (b.descendants.size == 1) {
				buildSynthDef.removeUGen(b);
			} {
				b.descendants.remove(this);
			};
			^(a - b.inputs[0])
		};

		if (a.isKindOf(UnaryOpUGen) and: { a.operator == 'neg' }) {
			// a.neg + b -> b - a
			if (a.descendants.size == 1) {
				buildSynthDef.removeUGen(a);
			} {
				a.descendants.remove(this);
			};
			^(b - a.inputs[0])
		};
		^nil
	}````

**Comment by telephon** [comment]
OK, here is a solution, but I am not sure if it is the best really. It seems a little strict that UGens have to return a graph that contains themselves. This goes against the idea of a SynthDef as a space into which UGens can insert themselves (rather than a tree of functions).

**Comment by danstowell** [comment]
Nice one Julian. Tricky. Certainly a lot of those optimisations assume that everything is in the resulting function graph (a graph though not necessarily a tree, I think). I've always thought of it that way myself too, though I see from your solution that there are quite common examples which don't fit that.

**Comment by telephon** [comment]
Yes, in SC1 as far as I know, you had to chain everything up. The reason we use SynthDefs (instead of just UGens) is that we can have this type of side effects.

I don't like my solution above for this reason. I've had another look, but also couldn't find the cause of the problem in ````optimizeAddNeg```` itself. Besides, only ````PureUGen````s should be subject to optimization really, and ````Poll```` and ````Out```` are not pure, they just add themselves to the tree and return something else (the input or zero respectively).

**Comment by mtmccrea** [comment]
When I run the original code Scott posted, it still looks like Poll is "optimize" out and doesn't poll the value:
```
(
d = SynthDef(\spreaderBroke, {
    arg steerDeg = 0.3;
    var steer, angB;

    steer = (steerDeg * 1.1).neg;
    steer.poll(label: \steer);
    angB = (steer.neg + 1.2).neg;
	Out.kr(0, steer);
}).add;
)

d.children // >> no Poll?

x = Synth(\spreaderBroke) // not polling...
```

I'm new to building SC, so it's possible I've done something wrong, but looks like I'm using the changes from this branch.

**Comment by scztt** [comment]
The problem seems to be: in the various optimizeAddNeg etc. cases, we remove a child without checking first if it's a PureUGen. Is the fix as simple as something like e.g.:
```
			if (b.isKindOf(PureUGen) && (b.descendants.size == 1)) {
				buildSynthDef.removeUGen(b);
			} {
				b.descendants.remove(this);
			};
```

**Comment by muellmusik** [comment]
Despite a lot of attention, this sort of thing seems to be very tricky to get right. Whatever happens with this, I think it would be a very good idea to add a SynthDef optimisation/sorting Unit Test to the CommonTests quark. I probably have some old test code that could be used for some of the problems that have come up in the past if that's of interest.

**Comment by scztt** [comment]
Oh.... I think I get it now. Our BinaryOpUGen optimizations are conditional on the number of descendants, but we're not *updating* the descendants when we change the graph. Here's a slightly simpler case:
```
d = SynthDef(\spreaderBroke, {
	var a, b, c, d, e;
	a = DC.kr(0.3);
	b = a + 1.1;
	c = b.neg;
	Poll.kr(1, c);
	
	d = 1.2 - c;
	e = d.abs;
}).add;

```
First optimization: since `(1.2 - c) == (1.2 - b.neg)`, we do `d = 1.2 + b`. Correct, but now `b` has a new dependent that isn't reflected in its dependents list.
Second optimization: since `(1.2 + b) == (1.2 +  a + 1.1)`, we can do `d = 1.2 + 1.1 + a`. Since it *looks* like b only has one dependent, we remove b from the graph completely. Now, we've effectively orphaned everything downstream of b (the Poll, in this example) - this will cause it to get purged later on when we do the final sort. It's NOT enough to just force the Poll to stick around because it's not a PureUGen, because it's input has already been removed from the graph.

I think the fix is to add `this.initTopoSort;` to the beginning of every `optimizeGraph` function. It's a bit brute-force, and will make synthdef compilation slower, but it's conservative and should guarantee the graph ends up alright.

**Comment by scztt** [comment]
@muellmusik - I think it would be reasonable to actually add a test to the SynthDef process that checks the list of non-PureUGens before and after optimization to make sure nothing is removed. Cost-wise it's trivial, and it would quickly highlight any optimization errors in the future (otherwise, it's fairly baffling to track down).

**Comment by muellmusik** [comment]
> Cost-wise it's trivial, and it would quickly highlight any optimization errors in the future (otherwise, it's fairly baffling to track down).

Hey @scztt, well, cue howls of protest at trivial cost! ;-) That would help of course, although the idealist in me would like it better if we understood the problem well enough that we could be confident in the implementation. :-\ 

IAC, removal of non-Pure UGens is one class of optimisation error. Incorrect reordering of WidthFirstUGens is another. I think a UnitTest that could be run before each release would at least ensure no regression of known problem cases, even if history cautions against total confidence!

**Comment by telephon** [comment]
Interesting.  @scztt, you mean to call ````UGen-initTopoSort```` at the beginning of every successfully selected optimization?  I am just a little worried, as synth def compilation time is a factor in many cases, and it might -- sometimes -- be better to accept unoptimized graphs.

**Comment by scztt** [comment]
@telephon - Actually, I think the most efficient-but-still-safe would be to call initTopoSort every time a UGen is either replaced or removed in optimization? Of course, more efficient than that would be to minimally patch up the dependencies inside of :replaceUGen instead re-initTopoSort-ing. But - we're already dangerous territory - if there's absolutely one thing SC *shoudn't* get wrong, it's SynthDef graphs. Until we can show some solid proof via analysis + extensive unit testing that an incremental update of the graph produces the same results as re-calling initTopoSort, I'd propose the re-sort plus some unit tests covering known fail cases and some common optimizations. I would feel safer if some of the optimization code could be refactored a bit - there's a lot of code duplication, and blurry boundaries (ugens directly modifying other ugens, rather than using e.g. methods on SynthDef or some other well-defined API). That at least would make it easier to make guarantees like "every optimization that modifies the graph happens in these N methods" etc. But - out of scope for 3.7, for sure. I'd love to use the rd_dot quark to visualize optimizations (i.e. removed nodes in red, replacements in blue), it would make debugging these things much easier.

I've got some very slow-to-build SynthDefs, I can benchmark to see what the performance hit is. I think there's a lot of room for optimization of initTopoSort that could make up some of the difference.

**Comment by telephon** [comment]
@scztt Yes, that makes a lot of sense: let's patch now with initTopoSort (unless it slows down things too much) and then improve the whole bit for 3.8

**Comment by vivid-synth** [comment]
This is all within sclang, correct? I.e. it's definitely "dead code elimination" gone amok, before the SynthDef is sent over the wire?

**Comment by scztt** [comment]
Yes, it's purely a bug in the graph optimization we do in sclang as we're building the def.

**Comment by jamshark70** [comment]
I was just looking at PureUGen and... correct me if this is a naïve reading, but wouldn't a correct algorithm be to mark the antecedents of a non-pure UGen as required, and not candidates for removal?

```
d = SynthDef(\spreaderBroke, {
    var a, b, c, d, e;
    a = DC.kr(0.3);  // 1.
    b = a + 1.1;  // 2.
    c = b.neg;  // 3.
    Poll.kr(1, c);  // 4.

    d = 1.2 - c;  // 5.
    e = d.abs;  // 6.
}).add;
```

1. `a` -- provisionally possible to optimize
2. `b` -- provisionally possible to optimize
3. `c` -- provisionally possible to optimize
4. Poll is non-pure, forcing `c` to be required. `c` depends on `b` which depends on `a`, so now all of 1-4 are mandatory.
5. `d` -- provisionally possible to optimize
6. `e` -- provisionally possible to optimize... and they never get plugged into any non-pure UGen, so they could be removed.

Then, *after* determining which pure UGens are and are not inputs of some nonpure UGen, then you can do the math optimizations. Maybe even drop `d` and `e` first -- if it's going to go away anyway, why are we doing math optimizations on them in the first place?

(I know 1-6 are not how it's working right now -- I'm asking if that would be a better approach: handle degenerate math cases on-the-fly, but withhold math optimizations until after checking pure UGen dependencies.)

**Comment by jamshark70** [comment]
E.g.

```
	performDeadCodeElimination {
		if(this.canRemove.not) { ^false };
		if (descendants.size == 0) {
			this.inputs.do {|a|
				if (a.isKindOf(UGen)) {
					a.descendants.remove(this);
					a.optimizeGraph
				}
			};
			buildSynthDef.removeUGen(this);
			^true;
		};
		^false
	}

	canRemove { ^false }
}

// ugen which has no side effect and can therefore be considered for a dead code elimination
// read access to buffers/busses are allowed

PureUGen : UGen {
	var <>canRemove = true;
```

Then UGen's init process marks canRemove = false for its inputs (all the way up the chains), and PureUGens init process skips that bit.

Untested and unfinished.

**Comment by jamshark70** [comment]
I played around with this a bit more this morning (starting 2018 off right!).

I don't think it will help to force a toposort repeatedly. The real problem is that the binary-op optimizations take a logic shortcut and sometimes optimize when it's incorrect to do so. If we fix that problem, then it would not be necessary to kill performance with repeated sorting.

### Logic

Focusing on `optimizeAddNeg` (since that seems to be central to this example): if you start with `1.2 - b.neg` and optimize to `1.2 + b`, we have these entities:

- `this`: the BinaryOpUGen `-`
  - `this.inputs[0]`: input 1.2
  - `this.inputs[1]`: input `UnaryOpUGen(\neg, b)`
- `optimized`: the BinaryOpUGen `+` [1.2, b]

When "optimized" replaces "this," then `this.inputs[1]` disappears from the graph and is no longer available to any other UGen. Therefore, it's incorrect to do the optimization if any input to the operator being replaced is also an input to any other unit.

If the optimization is valid, then both `this.inputs[0]` and `this.inputs[1]` will have exactly one descendant: `this`.

If either operand has two or more descendents, then that operand is needed somewhere else and you can't optimize it out.

### Bug

I created a branch with some debugging statements -- https://github.com/jamshark70/supercollider/tree/topic/synth-optimizations-debugging. This branch also includes my idea to use non-pure UGens to set a flag, simplifying dead code elimination.

Then:

```
(
UGen.testing = true;

d = SynthDef(\spreaderBroke, {
	var a, b, c, d, e;
	a = DC.kr(0.3);  // 1.
	b = a + 1.1;  // 2.
	c = b.neg;  // 3.
	Poll.kr(1, c);  // 4.
	
	d = 1.2 - c;  // 5.
	e = d.abs;  // 6.
	Out.kr(1000, e);
}).add;

d.children
)

->

before performDeadCodeElimination: [ a DC, a BinaryOpUGen, an UnaryOpUGen, an Impulse, a Poll, a BinaryOpUGen, an UnaryOpUGen, an Out ]
before optimizeGraph: [ a DC, a BinaryOpUGen, an UnaryOpUGen, an Impulse, a Poll, a BinaryOpUGen, an UnaryOpUGen, an Out ]
```

So far, so good: In this test case (modified from the original in this thread), everything plugs into a side effect, and nothing gets removed as dead code. (I have some other tests with dead code removal, where it works correctly.)

```
optimizeToSum3 checking: [ +, [ an OutputProxy, 1.1 ] ]
optimizeAddNeg: [ +, an OutputProxy, 1.1 ]
a: [ OutputProxy, nil, nil ]
b: [ Float, nil, nil ]
```

Again, so far, so good: checking the first `a + 1.1`, there's no second + or - to fold into Sum3, so, nothing happens, and there's no `neg` either.

```
optimizeSub: [ -, 1.2, an UnaryOpUGen ]
a: [ Float, nil, nil ]
b: [ UnaryOpUGen, neg, [ a Poll, a BinaryOpUGen ] ]
```

Now we have `1.2 - b.neg` -- but notice the second operand's descendents: the second operand is needed in Poll. Therefore this optimization *should* be suppressed.

BUT.

```
optimizeToSum3 checking: [ +, [ 1.2, a BinaryOpUGen ] ]
b: [ a BinaryOpUGen, +, [ an OutputProxy, 1.1 ], [ an UnaryOpUGen ] ]
```

It went ahead and did it anyway.

```
after optimizeGraph: [ a DC, nil, an UnaryOpUGen, an Impulse, a Poll, a Sum3, an UnaryOpUGen, an Out ]
reindexing: [ a DC, an UnaryOpUGen, an Impulse, a Poll, a Sum3, an UnaryOpUGen, an Out ]
-> [ a DC, a Sum3, an UnaryOpUGen, an Out, an Impulse ]
```

And then the toposort removes the Poll.

```
	optimizeSub {
		var a, b, replacement;
		#a, b = inputs;

		if (b.isKindOf(UnaryOpUGen) and: { b.operator == 'neg' }) {
			// a - b.neg -> a + b
			if (b.descendants.size == 1) {
				buildSynthDef.removeUGen(b);
			} {

// HERE'S THE BUG
// if b.descendants.size != 1 then WE MUST NOT PROCEED

				b.descendants.remove(this);
			};
			replacement = BinaryOpUGen('+', a, b.inputs[0]);

			synthDef.replaceUGen(this, replacement);
			replacement.optimizeGraph
		};
		^nil
	}
```

### Fix

In `optimizeSub`:

```
		if (b.isKindOf(UnaryOpUGen) and: { b.operator == 'neg' and: { b.descendants.size == 1 } }) {
			// a - b.neg -> a + b
			buildSynthDef.removeUGen(b);
			replacement = BinaryOpUGen('+', a, b.inputs[0]);

			synthDef.replaceUGen(this, replacement);
			replacement.optimizeGraph
		};
```

We should do the same in `optimizeAddNeg`.

I might not do it this morning, but we should check all math-op optimizations. It looks like the class library originally established the pattern of not checking the number of descendents, and later optimizations (Sum3 etc.) continued the pattern. But "go ahead with the optimization, no matter who else needs intermediate results" is logically not justifiable -- needs to be removed wherever it appears.

**Comment by jamshark70** [comment]
> we're not updating the descendants when we change the graph

Oh, just noticed, this is still the case. From the bug trace above:

```
optimizeToSum3 checking: [ +, [ 1.2, a BinaryOpUGen ] ]
b: [ a BinaryOpUGen, +, [ an OutputProxy, 1.1 ], [ an UnaryOpUGen ] ]
```

The UnaryOpUGen here is the `neg` that was (incorrectly) removed. So `optimizeSub` needs to set correct descendants too.

But I'm going to take a break now.

**Comment by muellmusik** [comment]
Hi James,

Two things:

Having swatted a few weird bugs in this stuff, I can tell you that this bit of the code can get pretty hairy pretty fast. In addition to PureUGens, you need to keep WidthFirstUGens in account. It would be a really good idea to get a few unit tests for SynthDef construction into place, so that we can check if any changes don’t introduce regressions and sort and optimise (or not) correctly. AFAICS these still don’t exist.

PureUGen’s subclasstree looks awfully short. Is that really correct?

S.

**Comment by muellmusik** [comment]
> PureUGen’s subclasstree looks awfully short. Is that really correct?

Ah sorry I see #3371 now.

> It would be a really good idea to get a few unit tests for SynthDef construction into place, so that we can check if any changes don’t introduce regressions and sort and optimise (or not) correctly.

Also do some benchmarks using a variety of different defs.

**Comment by jamshark70** [comment]
FWIW (and admitting it needs extensive testing), the top 2 commits here seem to fix the issue for which this issue was opened.

https://github.com/jamshark70/supercollider/tree/topic/synth-optimizations

```
d = SynthDef(\spreaderBroke, { 
    arg steerDeg = 0.3;
    var steer, angB;

    steer = (steerDeg * 1.1).neg;
    steer.poll(label: \steer);
    angB = (steer.neg + 1.2).neg;
    Out.kr(0, steer);
}).children;

-> [ a Control, a BinaryOpUGen, an UnaryOpUGen, a BinaryOpUGen, an UnaryOpUGen, an Out, an Impulse, a Poll ]
```

`steer.neg + 1.2` got correctly optimized into the second BinaryOpUGen. But the first `neg` is correctly *not* optimized out.

I did a few other tests too, all passed. But it's not extensive.

**Comment by muellmusik** [comment]
Can this be closed?

**Comment by jamshark70** [comment]
Should have been closed with #3437, I think.
